File: cpp/강철부대.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

vector<int> solution(int n, vector<vector<int>> roads, vector<int> sources, int destination) {
    vector<int> answer;

    vector<int> distance;
    vector<bool> visit;
    for (int i = 0; i < n + 1; i++) {
        distance.push_back(-1);
        visit.push_back(false);
    }

    vector<vector<bool>> adj;
    for (int i = 0; i < n + 1; i++) {
        vector<bool> row;
        for (int j = 0; j < n + 1; j++) {
            if (i == j) row.push_back(true);
            else row.push_back(false);
        }
        adj.push_back(row);
    }

    for (int i = 0; i < roads.size(); i++) {
        adj[roads[i][0]][roads[i][1]] = true;
        adj[roads[i][1]][roads[i][0]] = true;
    }



    queue<int> q;
    q.push(destination);
    visit[destination] = true;
    distance[destination] = 0;

    while (!q.empty()) {
        int location = q.front(); q.pop();

        for (int i = 1; i <= n; i++) {
            if (adj[location][i]) {
                if (!visit[i]) {
                    visit[i] = true;
                    q.push(i);
                    distance[i] = distance[location] + 1;
                }
            }
        }
    }

    for (int i = 0; i < sources.size(); i++) {
        answer.push_back(distance[sources[i]]);
    }

    return answer;
}

int main() {
    int n = 5;
    vector<vector<int>> roads = { {1, 2}, {1, 4}, {2, 4}, {2, 5}, {4, 5} };
    vector<int> sources = { 1, 3, 5 };
    int destination = 5;

    vector<int> result = solution(n, roads, sources, destination);

    for (int i = 0; i < result.size(); i++) {
        cout << result[i] << ' ';
    }
    cout << endl;
}

File: csharp/강철부대.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace CTwithCSharp
{
    class SteelMilitary
    {
        private Queue<int> q = null;    //BFS를 위한 큐
        bool[] visited = null;          //BFS를 위한 방문 여부를 저장하는 bool 배열
        int[] distance = null;          //BFS를 위해 목적지로부터 각 지역 사이의 거리를 저장하는 int 배열
        Dictionary<int, List<int>> road = null;  //roads 배열을 출발노드 : 도착노드리스트 형태로 저장
        const int MAX = 100001;         //최대값


        public int[] solution(int n, int[,] roads, int[] sources, int destination)
        {
            //초기화
            int[] answer = new int[sources.Length];
            visited = new bool[n + 1];
            distance = new int[n + 1];
            q = new Queue<int>();
            road = new Dictionary<int, List<int>>();
            for(int i = 0; i <= n; i++)
            {
                visited[i] = false;
                distance[i] = MAX;  //최대값으로 설정
                road[i] = new List<int>();
            }

            for(int i = 0; i < roads.GetLength(0); i++)
            {
                road[roads[i, 0]].Add(roads[i, 1]);
                road[roads[i, 1]].Add(roads[i, 0]);
            }

            q.Enqueue(destination);
            distance[destination] = 0;

            while(q.Count > 0)
            {
                int cur = q.Dequeue();

                if (visited[cur])  //이미 방문한 곳이라면 스킵
                    continue;

                visited[cur] = true;

                for(int i = 0; i < road[cur].Count; i++)
                {
                    q.Enqueue(road[cur][i]);
                    if (distance[road[cur][i]] > distance[cur] + 1)
                        distance[road[cur][i]] = distance[cur] + 1;
                }
            }

            for(int i = 0; i < sources.Length; i++)
            {
                if (distance[sources[i]] != MAX)
                    answer[i] = distance[sources[i]];
                else
                    answer[i] = -1;
            }


            return answer;
        }
    }
}
